src/network.rs

```diff
@@ -0,0 +1,151 @@
+    fn drop(&mut self) {
+        unsafe {
+            libc::freeifaddrs(self.buf);
+        }
+    }
+}
+
+#[cfg(any(target_os = "macos", target_os = "freebsd", target_os = "ios"))]
+impl From<&libc::sockaddr_dl> for MacAddr {
+    fn from(value: &libc::sockaddr_dl) -> Self {
+        let sdl_data = value.sdl_data;
+        // interface name length, NO trailing 0
+        let sdl_nlen = value.sdl_nlen as usize;
+        // make sure that it is never out of bound
+        if sdl_nlen + 5 < 12 {
+            MacAddr::from([
+                sdl_data[sdl_nlen] as u8,
+                sdl_data[sdl_nlen + 1] as u8,
+                sdl_data[sdl_nlen + 2] as u8,
+                sdl_data[sdl_nlen + 3] as u8,
+                sdl_data[sdl_nlen + 4] as u8,
+                sdl_data[sdl_nlen + 5] as u8,
+            ])
+        } else {
+            MacAddr::UNSPECIFIED
+        }
+    }
+}
+
+unsafe fn get_raw_ipv4(sock_addr: *const libc::sockaddr) -> u32 {
+    let sock_addr = sock_addr as *const libc::sockaddr_in;
+    (*sock_addr).sin_addr.s_addr
+}
+
+unsafe fn get_ipv4_interface_address(ifap: *const libc::ifaddrs) -> InterfaceAddress {
+    let sock_addr = (*ifap).ifa_addr;
+    let address = get_raw_ipv4(sock_addr);
+    let netmask = get_raw_ipv4((*ifap).ifa_netmask);
+
+    InterfaceAddress::IPv4(
+        Ipv4Addr::from(address.to_be()),
+        Ipv4Addr::from(netmask.to_be()),
+        // Ipv4Addr::from((address & netmask | (!netmask)).to_be())
+    )
+}
+
+#[cfg(any(target_os = "macos", target_os = "freebsd", target_os = "ios"))]
+unsafe fn parse_interface_address(ifap: *const libc::ifaddrs) -> InterfaceAddress {
+    let sock_addr = (*ifap).ifa_addr;
+    match (*sock_addr).sa_family as libc::c_int {
+        libc::AF_LINK => {
+            let addr = sock_addr as *const libc::sockaddr_dl;
+            InterfaceAddress::MAC(MacAddr::from(&*addr))
+        }
+        libc::AF_INET => get_ipv4_interface_address(ifap),
+        _ => InterfaceAddress::NotImplemented,
+    }
+}
+
+#[cfg(any(target_os = "linux", target_os = "android"))]
+unsafe fn parse_interface_address(ifap: *const libc::ifaddrs) -> InterfaceAddress {
+    use libc::sockaddr_ll;
+
+    let sock_addr = (*ifap).ifa_addr;
+    match (*sock_addr).sa_family as libc::c_int {
+        libc::AF_PACKET => {
+            let addr = sock_addr as *const sockaddr_ll;
+            // Take the first 6 bytes
+            let [addr @ .., _, _] = (*addr).sll_addr;
+            let addr = MacAddr::from(addr);
+            InterfaceAddress::MAC(addr)
+        }
+        libc::AF_INET => get_ipv4_interface_address(ifap),
+        _ => InterfaceAddress::NotImplemented,
+    }
+}
+
+/// Return an iterator on (interface_name, address) pairs
+pub(crate) fn get_interface_address() -> Result<InterfaceAddressIterator, String> {
+    let mut ifap = null_mut();
+    unsafe {
+        if libc::getifaddrs(&mut ifap) == 0 {
+            return Ok(InterfaceAddressIterator { ifap, buf: ifap });
```

Comment:
even if they're not supposed to be `NULL`, please check that `ifap` isn't `NULL` before creating the struct. Or wrap the fields into `NonNull` if you prefer.

Reply:
The null check is added👌
It is placed in `src/network_helper_nix.rs ` now.